Add VNET rules in table `azure_mysql_server`

Reference:

https://pkg.go.dev/github.com/Azure/azure-sdk-for-go@v58.0.0+incompatible/services/mysql/mgmt/2020-01-01/mysql#VirtualNetworkRulesClient.ListByServer

@ParthaI When we are creating any replica server list call is not working , it is just getting stuck.
Please find the below logs which may help.
```
request: StatusCode=500 -- Original Error: context canceled
2023-01-05 12:13:06.536 UTC [ERROR] steampipe-plugin-azure.plugin: [ERROR] 167292: table 'azure_mysql_server' failed to get column data, connectionCallId azure-167292: table 'azure_mysql_server' column 'server_configurations' requires hydrate data from listMySQLServersConfigurations, which failed with error mysql.ConfigurationsClient#ListByServer: Failure sending request: StatusCode=500 -- Original Error: context canceled.
2023-01-05 12:13:06.536 UTC [WARN]  steampipe-plugin-azure.plugin: [WARN]  167292: QueryData StreamError table 'azure_mysql_server' column 'server_configurations' requires hydrate data from listMySQLServersConfigurations, which failed with error mysql.ConfigurationsClient#ListByServer: Failure sending request: StatusCode=500 -- Original Error: context canceled.
```
<img width="356" alt="image" src="https://user-images.githubusercontent.com/46913995/210778032-08ec3cc8-768f-46ff-b0e5-e1b8819a8aee.png">

--- a/azure/table_azure_mysql_server.go
+++ b/azure/table_azure_mysql_server.go
@@ -207,6 +207,13 @@ func tableAzureMySQLServer(_ context.Context) *plugin.Table {
 				Hydrate:     listMySQLServersServerKeys,
 				Transform:   transform.FromValue(),
 			},
+			{
+				Name:        "vnet_rules",
+				Description: "The server keys of the server.",
+				Type:        proto.ColumnType_JSON,
+				Hydrate:     listMySQLServerVnetRules,
+				Transform:   transform.FromValue(),
+			},
 
 			// Steampipe standard columns
 			{

--- a/azure/table_azure_mysql_server.go
+++ b/azure/table_azure_mysql_server.go
@@ -209,7 +209,7 @@ func tableAzureMySQLServer(_ context.Context) *plugin.Table {
 			},
 			{
 				Name:        "vnet_rules",
-				Description: "The server keys of the server.",
+				Description: "VirtualNetworkRule a virtual network rule.",
 				Type:        proto.ColumnType_JSON,
 				Hydrate:     listMySQLServerVnetRules,
 				Transform:   transform.FromValue(),
@@ -387,9 +387,7 @@ func listMySQLServerVnetRules(ctx context.Context, d *plugin.QueryData, h *plugi
 
 	var vnetRules []mysql.VirtualNetworkRule
 
-	for _, i := range op.Values() {
-		vnetRules = append(vnetRules, i)
-	}
+	vnetRules = append(vnetRules, op.Values()...)
 
 	for op.NotDone() {
 		err = op.NextWithContext(ctx)
@@ -397,9 +395,7 @@ func listMySQLServerVnetRules(ctx context.Context, d *plugin.QueryData, h *plugi
 			plugin.Logger(ctx).Error("listMySQLServerVnetRules", "list_paging", err)
 			return nil, err
 		}
-		for _, i := range op.Values() {
-			vnetRules = append(vnetRules, i)
-		}
+		vnetRules = append(vnetRules, op.Values()...)
 	}
 
 	return vnetRules, nil

--- a/azure/table_azure_mysql_server.go
+++ b/azure/table_azure_mysql_server.go
@@ -390,7 +390,7 @@ func listMySQLServerVnetRules(ctx context.Context, d *plugin.QueryData, h *plugi
 	for op.NotDone() {
 		err = op.NextWithContext(ctx)
 		if err != nil {
-			plugin.Logger(ctx).Error("azure_mysql_server.listMySQLServerVnetRules", "api_error_pagging", err)
+			plugin.Logger(ctx).Error("azure_mysql_server.listMySQLServerVnetRules", "api_pagging_error", err)
 			return nil, err
 		}
 		vnetRules = append(vnetRules, op.Values()...)

--- a/azure/table_azure_mysql_server.go
+++ b/azure/table_azure_mysql_server.go
@@ -363,15 +363,13 @@ func listMySQLServersServerKeys(ctx context.Context, d *plugin.QueryData, h *plu
 }
 
 func listMySQLServerVnetRules(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateData) (interface{}, error) {
-	plugin.Logger(ctx).Trace("listMySQLServerVnetRules")
-
 	namespace := h.Item.(mysql.Server)
 	resourceGroup := strings.Split(string(*namespace.ID), "/")[4]
 	serverName := *namespace.Name
 
 	session, err := GetNewSession(ctx, d, "MANAGEMENT")
 	if err != nil {
-		plugin.Logger(ctx).Error("")
+		plugin.Logger(ctx).Error("azure_mysql_server.listMySQLServerVnetRules", "connection_error", err)
 		return nil, err
 	}
 	subscriptionID := session.SubscriptionID
@@ -381,7 +379,7 @@ func listMySQLServerVnetRules(ctx context.Context, d *plugin.QueryData, h *plugi
 
 	op, err := client.ListByServer(ctx, resourceGroup, serverName)
 	if err != nil {
-		plugin.Logger(ctx).Error("azure_mysql_server.listMySQLServerVnetRules", "list", err)
+		plugin.Logger(ctx).Error("azure_mysql_server.listMySQLServerVnetRules", "api_error", err)
 		return nil, err
 	}
 
@@ -392,7 +390,7 @@ func listMySQLServerVnetRules(ctx context.Context, d *plugin.QueryData, h *plugi
 	for op.NotDone() {
 		err = op.NextWithContext(ctx)
 		if err != nil {
-			plugin.Logger(ctx).Error("listMySQLServerVnetRules", "list_paging", err)
+			plugin.Logger(ctx).Error("azure_mysql_server.listMySQLServerVnetRules", "api_error_pagging", err)
 			return nil, err
 		}
 		vnetRules = append(vnetRules, op.Values()...)

--- a/azure/table_azure_mysql_server.go
+++ b/azure/table_azure_mysql_server.go
@@ -355,6 +355,49 @@ func listMySQLServersServerKeys(ctx context.Context, d *plugin.QueryData, h *plu
 	return mySQLServersServerKeys, nil
 }
 
+func listMySQLServerVnetRules(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateData) (interface{}, error) {
+	plugin.Logger(ctx).Trace("listMySQLServerVnetRules")
+
+	namespace := h.Item.(mysql.Server)
+	resourceGroup := strings.Split(string(*namespace.ID), "/")[4]
+	serverName := *namespace.Name
+
+	session, err := GetNewSession(ctx, d, "MANAGEMENT")
+	if err != nil {
+		plugin.Logger(ctx).Error("")
+		return nil, err
+	}
+	subscriptionID := session.SubscriptionID
+
+	client := mysql.NewVirtualNetworkRulesClient(subscriptionID)
+	client.Authorizer = session.Authorizer
+
+	op, err := client.ListByServer(ctx, resourceGroup, serverName)
+	if err != nil {
+		plugin.Logger(ctx).Error("azure_mysql_server.listMySQLServerVnetRules", "list", err)
+		return nil, err
+	}
+
+	var vnetRules []mysql.VirtualNetworkRule
+
+	for _, i := range op.Values() {
+		vnetRules = append(vnetRules, i)
+	}
+
+	for op.NotDone() {
+		err = op.NextWithContext(ctx)
+		if err != nil {
+			plugin.Logger(ctx).Error("listMySQLServerVnetRules", "list_paging", err)
+			return nil, err
+		}
+		for _, i := range op.Values() {
+			vnetRules = append(vnetRules, i)
+		}
+	}
+
+	return vnetRules, nil
+}
+
 func listMySQLServersConfigurations(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateData) (interface{}, error) {
 	plugin.Logger(ctx).Trace("listMySQLServersConfigurations")
 

--- a/docs/tables/azure_mysql_server.md
+++ b/docs/tables/azure_mysql_server.md
@@ -145,7 +145,7 @@ select
 from
   azure_mysql_server,
   jsonb_array_elements(server_configurations) as configurations
-where 
+where
    configurations ->> 'Name' = 'audit_log_enabled';
 ```
 
@@ -160,7 +160,7 @@ select
 from
   azure_mysql_server,
   jsonb_array_elements(server_configurations) as configurations
-where 
+where
    configurations ->'ConfigurationProperties' ->> 'value' = 'ON'
    and configurations ->> 'Name' = 'slow_query_log';
 ```
@@ -176,7 +176,20 @@ select
 from
   azure_mysql_server,
   jsonb_array_elements(server_configurations) as configurations
-where 
+where
    configurations ->'ConfigurationProperties' ->> 'value' = 'FILE'
    and configurations ->> 'Name' = 'log_output';
+```
+
+### Get VNET rules details of the server
+
+```sql
+select
+  name as server_name,
+  id as server_id,
+  rules -> 'properties' ->> 'ignoreMissingVnetServiceEndpoint' as ignore_missing_vnet_service_endpoint,
+  rules -> 'properties' ->> 'virtualNetworkSubnetId' as virtual_network_subnet_id
+from
+  azure_mysql_server,
+  jsonb_array_elements(vnet_rules) as rules;
 ```

